linked_list/linked_list.py

```diff
@@ -1,79 +1,159 @@
+            current_node = current_node.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        len_ll = 0
+        current_node = self.head
+
+        while current_node != None:
+            len_ll +=1
+            current_node = current_node.next
+        return len_ll
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        count = 0
+        current_node = self.head
+
+        while current_node != None:
+            if count == index:
+                return current_node.value
+            count += 1
+            current_node = current_node.next
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        head = self.head
+        if head == None:
+            return None
+
+        while head.next:
+            head = head.next
+        
+        return head.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.head == None:
+            self.add_first(value)
+        else:
+            new_node = Node(value)
+            current_node = self.head
+
+            while current_node.next:
+                current_node = current_node.next
+            current_node.next = new_node
+
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+        
+        head = self.head
+        max_value = self.get_at_index(0)
+
+        while head:
+            if head.value > max_value:
+                max_value = head.value
+            head = head.next
+
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(self, value):
-        pass
+
+        new = self.head
+
+        if new != None:
+            if new.value == value:
+                self.head = new.next
+                new = None
+                return
+
+            while new:
+                if new.value == value:
+                    break

```

Comment:
We generally try and avoid using break statements unless really necessary. Might you be able to refactor your code to eliminate this?